HW2/P5/driver.py

```diff
@@ -80,3 +104,16 @@ def randcolor():
             # SUBPROBLEM 3: sort objects by location.  Be sure to update the
             # grid if objects' indices change!  Also be sure to sort the
             # velocities with their object positions!
+
+            positions_sorted = zip(xrange(positions.shape[0]), (positions / grid_spacing).astype(int))
+            positions_sorted.sort(cmp_zorder)
+            indices = map(lambda x: x[0], positions_sorted)
+            velocities = velocities[indices]
+            positions = positions[indices]
+
+            # fix grid
+            for i in range(num_balls):
```

Comment:
for future reference, you can use a mask to do this really neatly in numpy without needing a for loop. For example, you can say something like

x_positions = positions[:, 0]
y_positions = positions[:, 1]
mask = (x_positions < 1) & (y_positions < 1) & (x_positions > 0) & (y_positions > 0)
grid[x_positions[mask], y_positions[mask] = np.arange(num_balls)[mask]